runtime/router_test.go

```diff
@@ -18,411 +18,201 @@
-			) {
-				var foo *string = nil
-				_ = *foo
-			},
-		).HandleRequest),
-	)
-	assert.NoError(t, err)
-
-	resp, err := gateway.MakeRequest("GET", "/panic", nil, nil)
-	if !assert.NoError(t, err) {
-		return
-	}
-
-	assert.Equal(t, resp.Status, "500 Internal Server Error")
-	assert.Equal(t, resp.StatusCode, 500)
-
-	bytes, err := ioutil.ReadAll(resp.Body)
-	if !assert.NoError(t, err) {
-		return
-	}
-
-	assert.Equal(t, bytes, []byte("Internal Server Error\n"))
-
-	allLogs := gateway.AllLogs()
-	assert.Equal(t, 1, len(allLogs))
+func (s *routerSuite) TestExtractParameters() {
+	err := s.router.Handle("GET", "/foo/:a", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		params := ParamsFromContext(r.Context())
+		s.Equal("x", params.Get("a"))
+		w.Write([]byte("foo\n"))
+	}))
+	s.NoError(err)
+	req := httptest.NewRequest("GET", "/foo/x", nil)
+	w := httptest.NewRecorder()
+	s.router.ServeHTTP(w, req)
+	s.Equal(http.StatusOK, w.Code)
+}
 
-	logLines := allLogs["A http request handler paniced"]
-	assert.Equal(t, 1, len(logLines))
+func (s *routerSuite) TestRouteConflict() {
+	err := s.router.Handle("GET", "/foo", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte("foo1\n"))
+	}))
+	s.NoError(err)
+
+	err = s.router.Handle("GET", "/foo", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte("foo2\n"))
+	}))
+	s.Error(err)
+
+	// Test that the original route is still registered and working correctly.
+
+	req := httptest.NewRequest("GET", "/foo", nil)
+	w := httptest.NewRecorder()
+	s.router.ServeHTTP(w, req)
+	s.Equal(http.StatusOK, w.Code)
+	body, _ := ioutil.ReadAll(w.Result().Body)
+	s.Equal([]byte("foo1\n"), body)
+}
 
-	line := logLines[0]
-	assert.Equal(t,
-		"wrapped: runtime error: "+
-			"invalid memory address or nil pointer dereference",
-		line["error"],
-	)
-	assert.Equal(t, "/panic", line["pathname"])
-	assert.Contains(
-		t,
-		line["errorVerbose"],
-		"runtime_test.TestRouterPanicNilPointer.func1",
-	)
+func (s *routerSuite) TestRouteConflictVariable() {
+	err := s.router.Handle("GET", "/foo/:a", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte("foo a\n"))
+	}))
+	s.NoError(err)
+
+	err = s.router.Handle("GET", "/foo/:b", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte("foo b\n"))
+	}))
+	s.Error(err)
+
+	// Test that the original route is still registered and working correctly.
+
+	req := httptest.NewRequest("GET", "/foo/x", nil)
+	w := httptest.NewRecorder()
+	s.router.ServeHTTP(w, req)
+	s.Equal(http.StatusOK, w.Code)
+	body, _ := ioutil.ReadAll(w.Result().Body)
+	s.Equal([]byte("foo a\n"), body)
 }
-func TestConflictingRoutes(t *testing.T) {
-	gateway, err := benchGateway.CreateGateway(
-		defaultTestConfig,
-		defaultTestOptions,
-		exampleGateway.CreateGateway,
-	)
-	if !assert.NoError(t, err) {
-		return
-	}
-	defer gateway.Close()
 
-	bgateway := gateway.(*benchGateway.BenchGateway)
+/*
```

Comment:
git is good at keeping history if you want to keep this one around. otherwise, let's delete.